File: CHTL/lexer/lexer.go
package lexer

import (
	"unicode"
)

// Lexer 词法分析器
type Lexer struct {
	input        string
	position     int
	readPosition int
	ch           byte
	line         int
	column       int
	keywords     map[string]TokenType
}

// NewLexer 创建新的词法分析器
func NewLexer(input string) *Lexer {
	l := &Lexer{
		input:    input,
		line:     1,
		column:   1,
		keywords: makeKeywords(),
	}
	l.readChar()
	return l
}

// makeKeywords 创建关键字映射
func makeKeywords() map[string]TokenType {
	return map[string]TokenType{
		"text":           KEYWORD_TEXT,
		"style":          KEYWORD_STYLE,
		"script":         KEYWORD_SCRIPT,
		"Template":       KEYWORD_TEMPLATE,
		"Custom":         KEYWORD_CUSTOM,
		"Origin":         KEYWORD_ORIGIN,
		"Import":         KEYWORD_IMPORT,
		"Namespace":      KEYWORD_NAMESPACE,
		"Configuration":  KEYWORD_CONFIGURATION,
		"use":            KEYWORD_USE,
		"inherit":        KEYWORD_INHERIT,
		"delete":         KEYWORD_DELETE,
		"insert":         KEYWORD_INSERT,
		"after":          KEYWORD_AFTER,
		"before":         KEYWORD_BEFORE,
		"replace":        KEYWORD_REPLACE,
		"at":             KEYWORD_AT_TOP,
		"top":            KEYWORD_AT_TOP,
		"bottom":         KEYWORD_AT_BOTTOM,
		"from":           KEYWORD_FROM,
		"as":             KEYWORD_AS,
		"except":         KEYWORD_EXCEPT,
		"html5":          KEYWORD_HTML5,
		"Var":            KEYWORD_VAR,
		"Element":        KEYWORD_ELEMENT,
		"Html":           KEYWORD_HTML,
		"JavaScript":     KEYWORD_JAVASCRIPT,
		"Chtl":           KEYWORD_CHTL,
		"CJmod":          KEYWORD_CJMOD,
		"Config":         KEYWORD_CONFIG,
		"vir":            KEYWORD_VIR,
		"listen":         KEYWORD_LISTEN,
		"delegate":       KEYWORD_DELEGATE,
		"animate":        KEYWORD_ANIMATE,
		"router":         KEYWORD_ROUTER,
		"util":           KEYWORD_UTIL,
		"then":           KEYWORD_THEN,
		"change":         KEYWORD_CHANGE,
		"fileloader":     KEYWORD_FILELOADER,
		"iNeverAway":     KEYWORD_INEVERAWAY,
		"printMylove":    KEYWORD_PRINTMYLOVE,
	}
}

// NextToken 获取下一个词法单元
func (l *Lexer) NextToken() Token {
	var tok Token

	l.skipWhitespace()

	tok.Position = Position{
		Line:   l.line,
		Column: l.column,
		Offset: l.position,
	}

	switch l.ch {
	case '{':
		tok = l.newToken(LEFT_BRACE, l.ch)
	case '}':
		tok = l.newToken(RIGHT_BRACE, l.ch)
	case '[':
		tok = l.newToken(LEFT_BRACKET, l.ch)
	case ']':
		tok = l.newToken(RIGHT_BRACKET, l.ch)
	case '(':
		tok = l.newToken(LEFT_PAREN, l.ch)
	case ')':
		tok = l.newToken(RIGHT_PAREN, l.ch)
	case ';':
		tok = l.newToken(SEMICOLON, l.ch)
	case ':':
		if l.peekChar() == '=' {
			l.readChar()
			tok = l.newToken(ASSIGN, "=")
		} else {
			tok = l.newToken(COLON, l.ch)
		}
	case ',':
		tok = l.newToken(COMMA, l.ch)
	case '.':
		tok = l.newToken(DOT, l.ch)
	case '=':
		if l.peekChar() == '=' {
			l.readChar()
			tok = l.newToken(EQUAL, "==")
		} else {
			tok = l.newToken(EQUALS, l.ch)
		}
	case '?':
		tok = l.newToken(QUESTION, l.ch)
	case '|':
		if l.peekChar() == '|' {
			l.readChar()
			tok = l.newToken(OR, "||")
		} else {
			tok = l.newToken(PIPE, l.ch)
		}
	case '&':
		if l.peekChar() == '&' {
			l.readChar()
			tok = l.newToken(AND, "&&")
		} else {
			tok = l.newToken(AMPERSAND, l.ch)
		}
	case '#':
		tok = l.newToken(HASH, l.ch)
	case '@':
		tok = l.newToken(AT, l.ch)
	case '$':
		tok = l.newToken(DOLLAR, l.ch)
	case '_':
		tok = l.newToken(UNDERSCORE, l.ch)
	case '+':
		tok = l.newToken(PLUS, l.ch)
	case '-':
		if l.peekChar() == '>' {
			l.readChar()
			tok = l.newToken(ARROW, "->")
		} else {
			tok = l.newToken(MINUS, l.ch)
		}
	case '*':
		if l.peekChar() == '*' {
			l.readChar()
			tok = l.newToken(POWER, "**")
		} else {
			tok = l.newToken(MULTIPLY, l.ch)
		}
	case '/':
		if l.peekChar() == '/' {
			tok = l.readLineComment()
		} else if l.peekChar() == '*' {
			tok = l.readBlockComment()
		} else {
			tok = l.newToken(DIVIDE, l.ch)
		}
	case '%':
		tok = l.newToken(MODULO, l.ch)
	case '<':
		if l.peekChar() == '=' {
			l.readChar()
			tok = l.newToken(LESS_EQUAL, "<=")
		} else {
			tok = l.newToken(LESS, l.ch)
		}
	case '>':
		if l.peekChar() == '=' {
			l.readChar()
			tok = l.newToken(GREATER_EQUAL, ">=")
		} else {
			tok = l.newToken(GREATER, l.ch)
		}
	case '!':
		if l.peekChar() == '=' {
			l.readChar()
			tok = l.newToken(NOT_EQUAL, "!=")
		} else {
			tok = l.newToken(NOT, l.ch)
		}
	case '"':
		tok = l.readString('"')
	case '\'':
		tok = l.readString('\'')
	case 0:
		tok = l.newToken(EOF, "")
	default:
		if isLetter(l.ch) {
			tok.Literal = l.readIdentifier()
			tok.Type = l.lookupIdent(tok.Literal)
			return tok
		} else if isDigit(l.ch) {
			tok = l.readNumber()
			return tok
		} else {
			tok = l.newToken(ILLEGAL, l.ch)
		}
	}

	l.readChar()
	return tok
}

// readChar 读取下一个字符
func (l *Lexer) readChar() {
	if l.readPosition >= len(l.input) {
		l.ch = 0
	} else {
		l.ch = l.input[l.readPosition]
	}
	l.position = l.readPosition
	l.readPosition++
	if l.ch == '\n' {
		l.line++
		l.column = 1
	} else {
		l.column++
	}
}

// peekChar 预览下一个字符
func (l *Lexer) peekChar() byte {
	if l.readPosition >= len(l.input) {
		return 0
	}
	return l.input[l.readPosition]
}

// readIdentifier 读取标识符
func (l *Lexer) readIdentifier() string {
	position := l.position
	for isLetter(l.ch) || isDigit(l.ch) || l.ch == '_' || l.ch == '-' {
		l.readChar()
	}
	return l.input[position:l.position]
}

// readNumber 读取数字
func (l *Lexer) readNumber() Token {
	position := l.position
	for isDigit(l.ch) || l.ch == '.' {
		l.readChar()
	}
	return l.newToken(NUMBER, l.input[position:l.position])
}

// readString 读取字符串
func (l *Lexer) readString(quote byte) Token {
	position := l.position + 1
	for {
		l.readChar()
		if l.ch == quote || l.ch == 0 {
			break
		}
	}
	return l.newToken(STRING, l.input[position:l.position])
}

// readLineComment 读取行注释
func (l *Lexer) readLineComment() Token {
	position := l.position
	for l.ch != '\n' && l.ch != 0 {
		l.readChar()
	}
	return l.newToken(COMMENT, l.input[position:l.position])
}

// readBlockComment 读取块注释
func (l *Lexer) readBlockComment() Token {
	position := l.position
	for {
		l.readChar()
		if l.ch == '*' && l.peekChar() == '/' {
			l.readChar()
			l.readChar()
			break
		}
		if l.ch == 0 {
			break
		}
	}
	return l.newToken(COMMENT, l.input[position:l.position])
}

// skipWhitespace 跳过空白字符
func (l *Lexer) skipWhitespace() {
	for l.ch == ' ' || l.ch == '\t' || l.ch == '\n' || l.ch == '\r' {
		l.readChar()
	}
}

// lookupIdent 查找标识符类型
func (l *Lexer) lookupIdent(ident string) TokenType {
	if tok, ok := l.keywords[ident]; ok {
		return tok
	}
	return IDENTIFIER
}

// newToken 创建新词法单元
func (l *Lexer) newToken(tokenType TokenType, ch interface{}) Token {
	var literal string
	switch v := ch.(type) {
	case byte:
		literal = string(v)
	case string:
		literal = v
	}
	return Token{
		Type:     tokenType,
		Literal:  literal,
		Position: Position{Line: l.line, Column: l.column, Offset: l.position},
	}
}

// isLetter 检查是否为字母
func isLetter(ch byte) bool {
	return unicode.IsLetter(rune(ch)) || ch == '_'
}

// isDigit 检查是否为数字
func isDigit(ch byte) bool {
	return '0' <= ch && ch <= '9'
}

File: chtl/src/lexer/lexer.cpp
#include "lexer/lexer.h"
#include <cctype>
#include <algorithm>

namespace chtl {
namespace lexer {

Lexer::Lexer(const std::string& source) 
    : source_(source), position_(0), line_(1), column_(1), start_position_(0) {
    initialize_keywords();
    initialize_operators();
}

void Lexer::initialize_keywords() {
    keywords_ = {
        // Basic keywords
        {"text", TokenType::TEXT},
        {"style", TokenType::STYLE},
        {"script", TokenType::SCRIPT},
        
        // Template keywords
        {"template", TokenType::TEMPLATE},
        {"@Style", TokenType::TEMPLATE_STYLE},
        {"@Element", TokenType::TEMPLATE_ELEMENT},
        {"@Var", TokenType::TEMPLATE_VAR},
        
        // Custom keywords
        {"Custom", TokenType::CUSTOM},
        {"@Style", TokenType::CUSTOM_STYLE},
        {"@Element", TokenType::CUSTOM_ELEMENT},
        {"@Var", TokenType::CUSTOM_VAR},
        
        // Origin keywords
        {"Origin", TokenType::ORIGIN},
        {"@Html", TokenType::ORIGIN_HTML},
        {"@Style", TokenType::ORIGIN_STYLE},
        {"@JavaScript", TokenType::ORIGIN_JAVASCRIPT},
        
        // Import keywords
        {"Import", TokenType::IMPORT},
        {"@Html", TokenType::IMPORT_HTML},
        {"@Style", TokenType::IMPORT_STYLE},
        {"@JavaScript", TokenType::IMPORT_JAVASCRIPT},
        {"@Chtl", TokenType::IMPORT_CHTL},
        {"@CJmod", TokenType::IMPORT_CJMOD},
        
        // Other keywords
        {"Namespace", TokenType::NAMESPACE},
        {"Configuration", TokenType::CONFIGURATION},
        {"use", TokenType::USE},
        {"except", TokenType::EXCEPT},
        {"delete", TokenType::DELETE},
        {"insert", TokenType::INSERT},
        {"inherit", TokenType::INHERIT},
        {"from", TokenType::FROM},
        {"as", TokenType::AS},
        {"at", TokenType::AT_TOP},
        {"top", TokenType::AT_TOP},
        {"bottom", TokenType::AT_BOTTOM},
        {"after", TokenType::AFTER},
        {"before", TokenType::BEFORE},
        {"replace", TokenType::REPLACE},
        
        // Boolean literals
        {"true", TokenType::BOOLEAN_LITERAL},
        {"false", TokenType::BOOLEAN_LITERAL}
    };
    
    initialize_chtl_keywords();
}

void Lexer::initialize_operators() {
    operators_ = {
        {"=", TokenType::ASSIGN},
        {":", TokenType::COLON},
        {";", TokenType::SEMICOLON},
        {",", TokenType::COMMA},
        {".", TokenType::DOT},
        {"->", TokenType::ARROW},
        {"+", TokenType::PLUS},
        {"-", TokenType::MINUS},
        {"*", TokenType::MULTIPLY},
        {"/", TokenType::DIVIDE},
        {"%", TokenType::MODULO},
        {"**", TokenType::POWER},
        {"==", TokenType::EQUAL},
        {"!=", TokenType::NOT_EQUAL},
        {"<", TokenType::LESS},
        {"<=", TokenType::LESS_EQUAL},
        {">", TokenType::GREATER},
        {">=", TokenType::GREATER_EQUAL},
        {"&&", TokenType::AND},
        {"||", TokenType::OR},
        {"!", TokenType::NOT},
        {"?", TokenType::QUESTION},
        {"(", TokenType::LEFT_PAREN},
        {")", TokenType::RIGHT_PAREN},
        {"[", TokenType::LEFT_BRACKET},
        {"]", TokenType::RIGHT_BRACKET},
        {"{", TokenType::LEFT_BRACE},
        {"}", TokenType::RIGHT_BRACE}
    };
}

void Lexer::initialize_chtl_keywords() {
    // CHTL特有的关键字和语法
    keywords_["html5"] = TokenType::IDENTIFIER; // HTML5类型
    keywords_["html"] = TokenType::IDENTIFIER;
    keywords_["head"] = TokenType::IDENTIFIER;
    keywords_["body"] = TokenType::IDENTIFIER;
    keywords_["div"] = TokenType::IDENTIFIER;
    keywords_["span"] = TokenType::IDENTIFIER;
    keywords_["p"] = TokenType::IDENTIFIER;
    keywords_["a"] = TokenType::IDENTIFIER;
    keywords_["img"] = TokenType::IDENTIFIER;
    keywords_["ul"] = TokenType::IDENTIFIER;
    keywords_["ol"] = TokenType::IDENTIFIER;
    keywords_["li"] = TokenType::IDENTIFIER;
    keywords_["table"] = TokenType::IDENTIFIER;
    keywords_["tr"] = TokenType::IDENTIFIER;
    keywords_["td"] = TokenType::IDENTIFIER;
    keywords_["th"] = TokenType::IDENTIFIER;
    keywords_["form"] = TokenType::IDENTIFIER;
    keywords_["input"] = TokenType::IDENTIFIER;
    keywords_["button"] = TokenType::IDENTIFIER;
    keywords_["title"] = TokenType::IDENTIFIER;
    keywords_["meta"] = TokenType::IDENTIFIER;
    keywords_["link"] = TokenType::IDENTIFIER;
    keywords_["script"] = TokenType::SCRIPT;
    keywords_["style"] = TokenType::STYLE;
    
    // CHTL JS关键字
    keywords_["fileloader"] = TokenType::IDENTIFIER;
    keywords_["listen"] = TokenType::IDENTIFIER;
    keywords_["delegate"] = TokenType::IDENTIFIER;
    keywords_["animate"] = TokenType::IDENTIFIER;
    keywords_["vir"] = TokenType::IDENTIFIER;
    keywords_["router"] = TokenType::IDENTIFIER;
    keywords_["iNeverAway"] = TokenType::IDENTIFIER;
    keywords_["util"] = TokenType::IDENTIFIER;
    keywords_["printMylove"] = TokenType::IDENTIFIER;
    
    // 配置关键字
    keywords_["INDEX_INITIAL_COUNT"] = TokenType::IDENTIFIER;
    keywords_["DEBUG_MODE"] = TokenType::IDENTIFIER;
    keywords_["DISABLE_NAME_GROUP"] = TokenType::IDENTIFIER;
    keywords_["DISABLE_STYLE_AUTO_ADD_CLASS"] = TokenType::IDENTIFIER;
    keywords_["DISABLE_STYLE_AUTO_ADD_ID"] = TokenType::IDENTIFIER;
    keywords_["DISABLE_DEFAULT_NAMESPACE"] = TokenType::IDENTIFIER;
    keywords_["DISABLE_CUSTOM_ORIGIN_TYPE"] = TokenType::IDENTIFIER;
    keywords_["DISABLE_SCRIPT_AUTO_ADD_CLASS"] = TokenType::IDENTIFIER;
    keywords_["DISABLE_SCRIPT_AUTO_ADD_ID"] = TokenType::IDENTIFIER;
    
    // 模块关键字
    keywords_["Info"] = TokenType::IDENTIFIER;
    keywords_["Export"] = TokenType::IDENTIFIER;
    keywords_["Name"] = TokenType::IDENTIFIER;
    keywords_["OriginType"] = TokenType::IDENTIFIER;
}

char Lexer::current_char() const {
    if (position_ >= source_.length()) {
        return '\0';
    }
    return source_[position_];
}

char Lexer::peek_char(size_t offset) const {
    size_t index = position_ + offset;
    if (index >= source_.length()) {
        return '\0';
    }
    return source_[index];
}

void Lexer::advance() {
    if (position_ < source_.length()) {
        if (source_[position_] == '\n') {
            line_++;
            column_ = 1;
        } else {
            column_++;
        }
        position_++;
    }
}

void Lexer::skip_whitespace() {
    while (position_ < source_.length() && is_whitespace(current_char())) {
        advance();
    }
}

void Lexer::skip_comment() {
    if (current_char() == '/' && peek_char() == '/') {
        skip_single_line_comment();
    } else if (current_char() == '/' && peek_char() == '*') {
        skip_multi_line_comment();
    } else if (current_char() == '-' && peek_char() == '-') {
        skip_generator_comment();
    }
}

void Lexer::skip_single_line_comment() {
    while (position_ < source_.length() && current_char() != '\n') {
        advance();
    }
}

void Lexer::skip_multi_line_comment() {
    advance(); // skip '/'
    advance(); // skip '*'
    
    while (position_ < source_.length()) {
        if (current_char() == '*' && peek_char() == '/') {
            advance(); // skip '*'
            advance(); // skip '/'
            break;
        }
        advance();
    }
}

void Lexer::skip_generator_comment() {
    while (position_ < source_.length() && current_char() != '\n') {
        advance();
    }
}

Token Lexer::scan_identifier() {
    start_position_ = position_;
    
    while (position_ < source_.length() && 
           (is_letter(current_char()) || is_digit(current_char()) || current_char() == '_')) {
        advance();
    }
    
    std::string value = source_.substr(start_position_, position_ - start_position_);
    
    // Check if it's a keyword
    auto it = keywords_.find(value);
    if (it != keywords_.end()) {
        return create_token(it->second, value);
    }
    
    return create_token(TokenType::IDENTIFIER, value);
}

Token Lexer::scan_string_literal() {
    start_position_ = position_;
    char quote = current_char();
    advance(); // skip opening quote
    
    while (position_ < source_.length() && current_char() != quote) {
        if (current_char() == '\\' && peek_char() == quote) {
            advance(); // skip backslash
        }
        advance();
    }
    
    if (position_ < source_.length()) {
        advance(); // skip closing quote
    }
    
    std::string value = source_.substr(start_position_, position_ - start_position_);
    return create_token(TokenType::STRING_LITERAL, value);
}

Token Lexer::scan_number_literal() {
    start_position_ = position_;
    
    while (position_ < source_.length() && 
           (is_digit(current_char()) || current_char() == '.')) {
        advance();
    }
    
    std::string value = source_.substr(start_position_, position_ - start_position_);
    return create_token(TokenType::NUMBER_LITERAL, value);
}

Token Lexer::scan_operator() {
    start_position_ = position_;
    
    // Check for multi-character operators first
    if (position_ + 1 < source_.length()) {
        std::string two_char = source_.substr(position_, 2);
        auto it = operators_.find(two_char);
        if (it != operators_.end()) {
            advance();
            advance();
            return create_token(it->second, two_char);
        }
    }
    
    // Single character operators
    std::string one_char(1, current_char());
    auto it = operators_.find(one_char);
    if (it != operators_.end()) {
        advance();
        return create_token(it->second, one_char);
    }
    
    return scan_unknown();
}

Token Lexer::scan_bracket() {
    char c = current_char();
    advance();
    
    switch (c) {
        case '(': return create_token(TokenType::LEFT_PAREN, "(");
        case ')': return create_token(TokenType::RIGHT_PAREN, ")");
        case '[': return create_token(TokenType::LEFT_BRACKET, "[");
        case ']': return create_token(TokenType::RIGHT_BRACKET, "]");
        case '{': return create_token(TokenType::LEFT_BRACE, "{");
        case '}': return create_token(TokenType::RIGHT_BRACE, "}");
        default: return scan_unknown();
    }
}

Token Lexer::scan_unknown() {
    char c = current_char();
    advance();
    return create_token(TokenType::UNKNOWN, std::string(1, c));
}

bool Lexer::is_letter(char c) const {
    return std::isalpha(c) || c == '_';
}

bool Lexer::is_digit(char c) const {
    return std::isdigit(c);
}

bool Lexer::is_whitespace(char c) const {
    return c == ' ' || c == '\t' || c == '\r' || c == '\n';
}

bool Lexer::is_operator_start(char c) const {
    return operators_.find(std::string(1, c)) != operators_.end();
}

bool Lexer::is_bracket(char c) const {
    return c == '(' || c == ')' || c == '[' || c == ']' || c == '{' || c == '}';
}

Token Lexer::create_token(TokenType type, const std::string& value) {
    return Token(type, value, line_, column_, start_position_);
}

Token Lexer::create_token(TokenType type) {
    return Token(type, "", line_, column_, start_position_);
}

TokenStream Lexer::tokenize() {
    std::vector<Token> tokens;
    
    while (position_ < source_.length()) {
        skip_whitespace();
        
        if (position_ >= source_.length()) {
            break;
        }
        
        char c = current_char();
        
        if (is_letter(c)) {
            tokens.push_back(scan_identifier());
        } else if (c == '"' || c == '\'') {
            tokens.push_back(scan_string_literal());
        } else if (is_digit(c)) {
            tokens.push_back(scan_number_literal());
        } else if (is_operator_start(c)) {
            tokens.push_back(scan_operator());
        } else if (is_bracket(c)) {
            tokens.push_back(scan_bracket());
        } else if (c == '/' && (peek_char() == '/' || peek_char() == '*')) {
            skip_comment();
        } else if (c == '-' && peek_char() == '-') {
            skip_comment();
        } else {
            tokens.push_back(scan_unknown());
        }
    }
    
    tokens.push_back(create_token(TokenType::EOF_TOKEN));
    return TokenStream(tokens);
}

std::vector<Token> Lexer::tokenize_to_vector() {
    return tokenize().get_tokens();
}

void Lexer::reset() {
    position_ = 0;
    line_ = 1;
    column_ = 1;
    start_position_ = 0;
}

} // namespace lexer
} // namespace chtl